map/generate_chart.cpp

```diff
@@ -0,0 +1,127 @@
+{
+  agg::rgba8 const kBackgroundColor(255, 255, 255, 255);
+  agg::rgba8 const kLineColor(0, 100, 100, 255);
+  double constexpr kLineWidth = 2;
+  uint32_t constexpr kBPP = 4;
+
+  if (width == 0 || height == 0 || geometry.empty())
+  {
+    ASSERT(false, ());
+    return false;
+  }
+
+  using TBlender = TBlendAdaptor<agg::rgba8, agg::order_rgba>;
+  using TPixelFormat = agg::pixfmt_custom_blend_rgba<TBlender, agg::rendering_buffer>;
+  using TBaseRenderer = agg::renderer_base<TPixelFormat>;
+  using TPrimitivesRenderer = agg::renderer_primitives<TBaseRenderer>;
+  using TSolidRenderer = agg::renderer_scanline_aa_solid<TBaseRenderer>;
+
+  agg::rendering_buffer renderBuffer;
+  TPixelFormat pixelFormat(renderBuffer, agg::comp_op_src_over);
+  TBaseRenderer m_baseRenderer(pixelFormat);
+
+  frameBuffer.resize(width * kBPP * height);
+  memset(&frameBuffer[0], 0, frameBuffer.size());
+  renderBuffer.attach(&frameBuffer[0], static_cast<unsigned int>(width),
+                      static_cast<unsigned int>(height),
+                      static_cast<unsigned int>(width * kBPP));
+  m_baseRenderer.reset_clipping(true);
+  unsigned op = pixelFormat.comp_op();
+  pixelFormat.comp_op(agg::comp_op_src);
+  m_baseRenderer.clear(kBackgroundColor);
+  pixelFormat.comp_op(op);
+
+  agg::rasterizer_scanline_aa<> rasterizer;
+  rasterizer.clip_box(0, 0, width, height);
+  typedef agg::poly_container_adaptor<vector<m2::PointD>> path_t;
+  path_t path_adaptor(geometry, false);
+  typedef agg::conv_stroke<path_t> stroke_t;
+
+  stroke_t stroke(path_adaptor);
+  stroke.width(kLineWidth);
+  stroke.line_cap(agg::round_cap);
+  stroke.line_join(agg::round_join);
+  rasterizer.add_path(stroke);
+
+  agg::scanline32_p8 scanline;
+  agg::render_scanlines_aa_solid(rasterizer, scanline, m_baseRenderer, kLineColor);
+
+//  vector<uint8_t> output;
+//  il::EncodePngToMemory(width, height, frameBuffer, output);
+//  ofstream file("/Users/vladimirbykoyanko/image.png");
+//  file.write(reinterpret_cast<char *>(output.data()), output.size());
+  return true;
+}
+
+bool GenerateChart(size_t width, size_t height, feature::TAltitudes const & yAxisNormalizedData,
+                   vector<uint8_t> & frameBuffer)
+{
+  uint32_t constexpr kHeightIndent = 10;
+
+  auto const minMaxAltitudeIt = minmax_element(yAxisNormalizedData.begin(), yAxisNormalizedData.end());
+  double const minAlt = *minMaxAltitudeIt.first;
+  double const maxAlt = *minMaxAltitudeIt.second;
+  double const deltaAlt = maxAlt - minAlt;
+  double const oneSegLenPix = static_cast<double>(width) / yAxisNormalizedData.size();
+
+  vector<m2::PointD> geometry(yAxisNormalizedData.size());
+  for (size_t i = 0; i < yAxisNormalizedData.size(); ++i)
+  {
+    geometry[i] = m2::PointD(i * oneSegLenPix,
+                             kHeightIndent + static_cast<double>(height - 2 * kHeightIndent) * (yAxisNormalizedData[i] - minAlt) /
+                             deltaAlt);
+    geometry[i].y = height - geometry[i].y;
```

Comment:
Why not to include "height - " to the formula that computes geometry[i] for the first time?